File: lib/src/database/database_expense.dart
import 'package:cents/src/domain/amount.dart';
import 'package:cents/src/domain/expense.dart';
import 'package:cents/src/domain/expense_category.dart';
import 'package:floor/floor.dart';

@Entity(tableName: "Expenses")
class DatabaseExpense {
  @PrimaryKey(autoGenerate: true)
  final int id;
  final Amount cost;
  final ExpenseCategory category;
  final DateTime createdAt;
  final String note;

  DatabaseExpense(this.id, this.cost, this.category, this.createdAt, this.note);

  DatabaseExpense.fromDomain(Expense expense)
      : this(expense.id, expense.cost, expense.category, expense.createdAt,
            expense.note);

  Expense asDomain() {
    return Expense(
        id: id,
        cost: cost,
        category: category,
        createdAt: createdAt,
        note: note);
  }
}

File: app/src/main/java/jaanc/cents/database/DatabaseExpense.kt
package jaanc.cents.database

import androidx.room.Entity
import androidx.room.PrimaryKey
import jaanc.cents.domain.Amount
import jaanc.cents.domain.Expense
import jaanc.cents.domain.ExpenseCategory
import java.time.LocalDateTime

@Entity(tableName = "Expenses")
data class DatabaseExpense(
    @PrimaryKey(autoGenerate = true) val id: Int,
    val cost: Amount,
    val category: ExpenseCategory,
    val note: String,
    val createdAt: LocalDateTime,
) {
    companion object {
        fun of(expense: Expense): DatabaseExpense {
            return DatabaseExpense(
                expense.id,
                expense.cost,
                expense.category,
                expense.note,
                expense.createdAt
            )
        }
    }

    fun asDomain(): Expense {
        return Expense(id, category, cost, note, createdAt)
    }
}